instance/tests/models/test_openedx_instance.py

```diff
@@ -241,3 +245,31 @@ def test_spawn_appserver_with_external_databases(self, mocks, mock_provision):
                         'EDXAPP_MONGO_USER', 'EDXAPP_MONGO_PASSWORD',
                         'EDXAPP_SWIFT_USERNAME', 'EDXAPP_SWIFT_KEY'):
             self.assertTrue(ansible_vars[setting])
+
+    @ddt.data(True, False)
+    @patch_services
+    @patch('instance.models.openedx_instance.OpenEdXAppServer.terminate_vm')
+    def test_delete_instance(self, mocks, delete_by_ref, mock_terminate_vm):
+        """
+        Test that an instance can be deleted directly or by its InstanceReference, and that the
+        associated AppServers and VMs will be terminated.
+        """
+        instance = OpenEdXInstanceFactory(sub_domain='test.deletion', use_ephemeral_databases=False)
+        instance_ref = instance.ref
+        appserver = OpenEdXAppServer.objects.get(pk=instance.spawn_appserver())
+
+        self.assertEqual(mock_terminate_vm.call_count, 0)
+
+        # Now delete the instance, either using InstanceReference or the OpenEdXInstance class:
+        if delete_by_ref:
+            instance_ref.delete()
+        else:
+            instance.delete()
+
+        self.assertEqual(mock_terminate_vm.call_count, 1)
+        with self.assertRaises(OpenEdXInstance.DoesNotExist):
+            OpenEdXInstance.objects.get(pk=instance.pk)
```

Comment:
@itsjeyd I did that at first, but `OpenEdXInstance` overrides `refresh_from_db()` to also refresh the `InstanceReference` at the same time, so that ended up generating the `InstanceReference.DoesNotExist` exception instead of the `OpenEdXInstance.DoesNotExist` that I wanted to test here. So I have to do it this way to be more direct.